Implement material iterative prestressing

amgebauer opened the pull request:
This adds a general material wrapper to prestress any material with iterative prestressing.

In collaboration with @datzj and our student Lena Heiß.

datzj commented on src/mat/4C_mat_iterative_prestress.cpp:
```diff
@@ -0,0 +1,326 @@
+    // convert original stress to stress like Voigt notation
+    Core::LinAlg::Voigt::Stresses::vector_to_matrix(stress_elastic, S_elastic);
+
+    // comute G*S_tilde*GT
```
```suggestion
    // compute G*S_tilde*GT
```

datzj commented on src/mat/4C_mat_iterative_prestress.cpp:
```diff
@@ -0,0 +1,326 @@
+    GS_tilde.multiply_nn(prestretch_tensor, S_elastic);
+    GS_tildeGT.multiply_nt(GS_tilde, prestretch_tensor);
+
+    // comute pk2 in reference configuration
```
```suggestion
    // compute pk2 in reference configuration
```

datzj commented on src/inpar/4C_inpar_validmaterials.cpp:
```diff
@@ -3596,6 +3596,19 @@ std::shared_ptr<std::vector<std::shared_ptr<Mat::MaterialDefinition>>> Input::va
+        Core::Materials::m_iterative_prestress);
+
+    add_named_int(m, "MATID", "Id of the material");
+    add_named_bool(m, "ACTIVE", "True if prestress should be updated");
```
I don't know if the parameter description is fully clear.

datzj commented on src/mat/4C_mat_iterative_prestress.hpp:
```diff
@@ -0,0 +1,125 @@
+
+FOUR_C_NAMESPACE_CLOSE
+
+#endif  // !BACI_MAT_ITERATIVE_PRESTRESS_HPP
```
```suggestion
#endif  // !FOUR_C_MAT_ITERATIVE_PRESTRESS_HPP
```

datzj reviewed (dismissed):
Thank you for preparing this feature to be merged! I only added a few minor comments/typo fixes. I think the functionality was tested thoroughly in the term paper and also in the test you added.

amgebauer commented:
@datzj Thanks for spotting the typos. I fixed them.

datzj reviewed (dismissed):
LGTM, thanks!

lauraengelhardt reviewed (commented):
I left two suggestions/questions 😊 Otherwise everything looks good!

amgebauer commented:
Thanks for your review @lauraengelhardt. I added it to the mat_service. On the way, I fixed an error in the pull_back operation and reused it in another material that had its own implementation.

Now ready for a second round of reviews.

lauraengelhardt reviewed (approved):
Look's good, thanks for incorporating the suggestions!

datzj reviewed (approved):
Looks good!
